src/Lecture3.hs

```diff
@@ -178,7 +222,17 @@ together only different elements.
 Product {getProduct = 6}
 
 -}
-appendDiff3 = error "TODO"
+
+appendDiff3 :: (Eq a, Semigroup a) => a -> a -> a -> a
+appendDiff3 x y z
+  | z /= x && z /= y = appendDiff2 (appendDiff2 x y) z
+  | otherwise        = appendDiff2 x y 
+
+-- despite ScopedTypeVariables, looks like we can still use them as value variables
+appendDiff2 :: (Eq a, Semigroup a) => a -> a -> a
+appendDiff2 a b
+  | a == b    = a
+  | otherwise = a <> b
```

Comment:
`ScopedTypeVariables` only brings type to the scope. The variables `a` and `b` in the `appendDiff2` function body are term-level values. So no conflict here 🙂